core/src/ldml/ldml_processor.cpp

```diff
@@ -519,14 +524,40 @@ ldml_event_state::ldml_event_state(
   this->modifier_state = m;
   this->is_key_down    = i;
   this->event_flags    = e;
+
+  actions.persist_options = new km_core_option_item[1];
+  actions.persist_options[0] = NULL_OPTIONS[0];
+
+  // initialize actions
+  clear();
 }
 
 void ldml_event_state::commit() {
-  state->actions().commit();
+  // TODO-LDML: is this necessary? what's the lifecycle of actions.output?
+  actions.output = new km_core_usv[text.length()+1];
+  memcpy(actions.output, text.c_str(), text.length()+1);
```

Comment:
memcpy isn't right here -- text.length() is in chars but we are looking at USVs. Something like:

```suggestion
  memcpy(actions.output, text.c_str(), (text.length()+1) * sizeof(text[0]));
```

Reply:
Ooh thanks.  I still wonder what the lifecycle is. If it's guaranteed to not be used past the return, i could just pass in `text.c_str()`.

Reply:
Yeah, we would always copy it internally so that ownership of the struct is clear. At present `set_actions` just pulls the data out of the struct and sticks it in the queue.

The owner should be able to pass a valid action struct to `km_core_actions_dispose()`. And `set_actions` does not transfer ownership.

Reply:
can that string be `const km_core_usv* output;` then?

Reply:
hmm, I _think_ so?